CMakeLists.txt

```diff
@@ -421,7 +427,18 @@ if(MSVC)
     add_definitions(-D_CRT_NONSTDC_NO_DEPRECATE)
 endif()
 
-if(BASEARCH_X86_FOUND)
+if(BASEARCH_PPC_FOUND)
+    # Check if we have what we need for POWER8 optimizations
+    set(CMAKE_REQUIRED_FLAGS "${POWER8FLAG}")
```

Comment:
I think you need the `if(WITH_NATIVE_INSTRUCTIONS)` statement found in `X86_FOUND`..

Reply:
@nmoinvaz I don't believe the extra checks as done for x86 are necessary. The feature flags are already overwritten if `WITH_NATIVE_INSTRUCTIONS` is enabled:

https://github.com/zlib-ng/zlib-ng/blob/2ded41c49e1efb1b0693649a70d52c8697267481/CMakeLists.txt#L248-L253

So at that section you pointed (after it, actually. on the x86 portion), each flag will already have the proper value, the extra checks do not seem necessary even for the x86 case.

But I'm a bit confused as to what this `WITH_NATIVE_INSTRUCTIONS` feature should do. When this is set on `CMakeLists.txt`, all feature flags are overwritten (snippet above), so all later feature checks are done using this flag instead. But on `configure` with `--native` (which should be equivalent), the feature flags are not changed, only `-march=native` is appended to CFLAGS, so files seem to compiled with different flags depending on the build system.

Reply:
Yes you are correct, seems unnecessary for X86 as you say. But I think u have to override `POWER8FLAG` with `NATIVEFLAG` too correct?

Reply:
Result of -march=native is that even the files that don't use any vector intrinsics or instructions can be optimized by compiler using vector instructions... Without it, those files are limited to general-purpose registers.

Reply:
Yes, that can be done. But I don't see a way to do the same for the `configure` case, as the `POWER8FLAG`  flag is defined in `arch/power/Makefile.in` (`P8FLAGS` there). Looks like that is not done for other archs either, but I'd like to avoid that if possible.

One (but maybe hacky) way to fix this would be to use `sed` to replace the flag on `arch/power/Makefile.in` if `--native` is enabled.

Reply:
What CMakeLists.txt is completely incorrect... It should add NATIVEFLAG to CFLAGS like configure does...

Reply:
In that case maybe X86 checks should look like `set(CMAKE_REQUIRED_FLAGS "${SSE4FLAG} ${NATIVEFLAG}")

Reply:
Ok, so in that case I'll simply add some code to properly setup the `NATIVEFLAG` for POWER (should be `-mcpu=native` on POWER according to [GCC docs](https://gcc.gnu.org/onlinedocs/gcc-10.1.0/gcc/RS_002f6000-and-PowerPC-Options.html#RS_002f6000-and-PowerPC-Options). I couldn't find on clang docs, but testing shows it's the same case). And I'll let the `POWER8FLAG` stay the way it is, to guarantee the optimized code is compiled with the proper `-mcpu` flag. Is that fine?

Reply:
Sounds good.

Reply:
Just to clarify this for everyone.
-march=native means gcc gets to check what CPU it is currently running on and will enable all instruction sets (and set cache parameters in regards to prefetching) to match the cpu. This means the compiled binaries cannot be run on any cpu without all the instruction sets of the cpu that was used to compile it.
But: what if the compiling cpu is say a sandy lake cpu? That lacks AVX2, so AVX2 code would then fail to build unless you also pass the -mavx2 parameter for the files utilizing those intrinsics.

So, yes. -march=native will be optimal for the current cpu (useless for distro packages though), but we still need the instruction set enablers appended after that to make sure all the appropriate flags are enabled.

configure gets this right, cmake still has a way to go in order to clean this up, and setting the correct flags per-file.

Reply:
The way cmake does it now limits it to instructions the current CPU can handle unless cross-compiling. This is partly because it tries to run the test binaries, not just compile and because previously we did set the flags for all files, not just the ones that need them.